### ISSUE TITLE
Some circular imports should be reported as an error.

### ISSUE THREAD

**Issue by mrolle45** [python/mypy#12491]
I have two modules:
```
A.py:
from B import X
Y = 2

B.py:
from A import Y
X = 1
```
Importing either module will raise an `ImportError`.  Importing `A` first will import `B`, which will fail on `from A import Y`.  Similarly if `B` is imported first.
So this is clearly wrong code, but `mypy` does not report this as an error!

I realize that mypy is a static type checker, not a dynamic type checker, and so it does not execute any code.  However, it ought to catch programming errors which could lead to exceptions being raised or uncertain results.

Here's my view on import dependencies.  There are two kinds, which I am calling **immediate** and **delayed**. An immediate dependency means that when `A.py` is imported, this causes `B.py` to be imported before the `A` code has completed execution.  A runtime dependency means that `B.py` is imported (possibly) some time after `A.py` has completely executed.
A circular dependency is when there is a cycle of immediate dependencies.

In `build.py`, there is a notion of a top-level import statement.  This is implemented as saying whether the statement is an element of the module's outer block, rather than being contained in a class or function scope.

I assert that this is incorrect, in that it misses one type of immediate dependency.  That is, an import which is contained in a class def (which may be in another class def, and so on).  The reason is that all code in such classes is executed as part of the import, and in the order such statements appear in the module code as a whole.  The effect on the sequence of imports is identical to what it would be if these nested imports had appeared at the top level.  The one difference is that the objects created by the import statement are bound into the local scope, not the global scope.  (Note that any import contained somewhere within a function def will _not_ be executed at import time.)

Think of a "pending imports" stack, initially empty.  At any given time, there can be several (different) modules which are temporarily suspended at the point where they immediately import the next one.  For each module, we have (1) its current partially built state at the point of the import of the next module, and (2) the trailing code which follows that import.  The last module is currently running.
Because of the cycle of immediate imports, every module in the cycle will find its way onto the stack.  After that, the last module will run to completion, followed by the entire trailing code of the other modules, in reverse order.
If a module `A` references a name `B.X` (both modules being part of the cycle), then the latter will refer to either:
1. `B.X` in the partially built state of `B`, if `B` is earlier on the stack than `A`, or
2. `B.X` in the final state of `B`, if `B` is later on the stack than `A`.

The problem occurs in case 1, if `B.X` is assigned (or perhaps reassigned) in the `B` trailing code.  The reference in `A` to `B.X` will be incorrect, or more likely, undefined (causing an `ImportError` at runtime).
If `A` has `from B import *`, then mypy needs to determine which members of `B` would be imported from a completely built `B`, and if any of these are set in the trailing code, that would be an error.

I propose the following algorithm to implement this error detection:
1. Consider every import statement as **immediate** if it is contained in the top level or in an **immediate** class def.  A class def is **immediate** by the same criterion.  Unreachable or MYPY protected code is exempted.
2. Every immediate import has either `PRI_HIGH` or `PRI_MED` priority.  `PRI_MED` is only for `import B`, provided that there is no reference to any name `B.X` in any immediate code.  Any such reference before the `import B` would be an error.
3. Form SCCs (strongly connected components) of modules using these two dependency priorities.  This is done in existing code using these and all lower priorities.  Break these up into SCCs for just  `PRI_HIGH` or `PRI_MED`.
4. For each SCC, do the following:
5. For each module `A` in the SCC, determine the pending stack, with `A` being imported first.
6. For each module `B` in the pending stack, look for references to any `C.X` where `C` precedes `B` in the stack.  All such references from `B` to any other module in the cycle can be collected once, along with the line number where the member was (last) assigned.
7. If `C.X` is last assigned after the point where `C` is suspended, then this is the error we are looking for.
8. The error report should cite the reference to `C.X` in `B` and a note that the result of the reference can vary with which module in the SCC is first imported.
9. I suggest that these error reports be issued only once each, and that the entire SCC processing be followed by notes listing the modules in the SCC.

**Comment by JelleZijlstra** [comment]
Note that pyright implements a check of this kind. (Not sure if it works exactly like you propose.)
